File: src/main/java/engine/TTable.java
/*
 * Copyright (c) Joseph Prichard 2023.
 */

package engine;

import lombok.Getter;

import javax.annotation.Nullable;
import java.util.Random;

public class TTable {

    public record Node(long key, float heuristic, int depth) {
    }

    private final int[][] table;
    private final Node[][] cache;
    @Getter
    private int hits = 0;
    @Getter
    private int misses = 0;

    public TTable(int tableSize) {
        // each cache line has 2 elements, one being "replace by depth" and one being "replace always"
        this.cache = new Node[tableSize][2];

        int tableLen = OthelloBoard.getBoardSize() * OthelloBoard.getBoardSize();
        Random generator = new Random();
        table = new int[tableLen][3];
        for (int i = 0; i < tableLen; i++) {
            for (int j = 0; j < 3; j++) {
                int n = generator.nextInt();
                table[i][j] = n >= 0 ? n : -n;
            }
        }
    }

    public long hash(OthelloBoard board) {
        long hash = 0;
        for (int i = 0; i < table.length; i++) {
            hash = hash ^ table[i][board.getSquare(i)];
        }
        return hash;
    }

    public void clear() {
        for (Node[] node : cache) {
            node[0] = null;
            node[1] = null;
        }
    }

    public void put(Node node) {
        int h = (int) (node.key() % cache.length);
        Node[] cacheLine = cache[h];
        // check if "replace by depth" is populated
        if (cacheLine[0] != null) {
            // populated, new node is better so we do replacement
            if (node.depth() > cacheLine[0].depth()) {
                cacheLine[1] = cacheLine[0];
                cacheLine[0] = node;
            } else {
                // new node is worse, so it should be sent to "replace always"
                cacheLine[1] = node;
            }
        } else {
            // not populated, so it can be used
            cacheLine[0] = node;
        }
    }

    @Nullable
    public Node get(long key) {
        int h = (int) (key % cache.length);
        Node[] cacheLine = cache[h];

        for (Node n : cacheLine) {
            if (n != null && n.key() == key) {
                hits++;
                return n;
            }
        }
        misses++;
        return null;
    }
}

File: app/othello/ttable.go
package othello

import (
	"math/rand"
	"time"
)

type Node struct {
	Set       bool
	Key       uint64
	Board     Board
	Heuristic float64
	Depth     int
}

type TTable struct {
	table  [][3]int
	cache  [][2]Node
	hits   int
	misses int
}

func NewTTable(cacheSize int, getRand func() int) TTable {
	cache := make([][2]Node, cacheSize)

	tableLen := BoardSize * BoardSize
	table := make([][3]int, tableLen)

	for i := 0; i < tableLen; i++ {
		for j := 0; j < 3; j++ {
			n := getRand()
			table[i][j] = n
		}
	}
	return TTable{table: table, cache: cache}
}

func NewTTableWithRand(cacheSize int) TTable {
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	getRand := func() int {
		return r.Intn(cacheSize)
	}
	return NewTTable(cacheSize, getRand)
}

func (tt *TTable) Hash(board Board) uint64 {
	var hash uint64 = 0
	for i := 0; i < len(tt.table); i++ {
		square := board.GetSquareByPosition(i)
		hash ^= uint64(tt.table[i][square])
	}
	return hash
}

func (tt *TTable) Clear() {
	tt.hits = 0
	tt.misses = 0
	for i := range tt.cache {
		tt.cache[i][0] = Node{}
		tt.cache[i][1] = Node{}
	}
}

func (tt *TTable) Put(key uint64, board Board, heuristic float64, depth int) {
	node := Node{
		Key:       key,
		Board:     board,
		Heuristic: heuristic,
		Depth:     depth,
		Set:       true,
	}

	h := int(node.Key % uint64(len(tt.cache)))
	cacheLine := &tt.cache[h]

	if cacheLine[0].Set {
		if node.Depth > cacheLine[0].Depth {
			cacheLine[1] = cacheLine[0]
			cacheLine[0] = node
		} else {
			cacheLine[1] = node
		}
	} else {
		cacheLine[0] = node
	}
}

func (tt *TTable) Get(key uint64, board Board) (Node, bool) {
	h := int(key % uint64(len(tt.cache)))

	cacheLine := &tt.cache[h]
	for _, n := range cacheLine {
		if n.Set && n.Board == board {
			tt.hits++
			return n, true
		}
	}
	tt.misses++
	return Node{}, false
}

func (tt *TTable) Hits() int {
	return tt.hits
}

func (tt *TTable) Misses() int {
	return tt.misses
}
